python/ray/tune/search/placeholder.py

```diff
@@ -0,0 +1,124 @@
+from typing import Any, Dict, Tuple
+from ray.tune.search.sample import Categorical, Function
+from ray.tune.search.variant_generator import assign_value, _get_value
+
+from ray.util.annotations import DeveloperAPI
+
+
+class _Categorical_Value:
+    """Replaced value for categorical typed objects.
+    """
+    def __init__(self, choices):
+        self._choices = choices
+
+    def get(self, ph):
+        ca, i = ph.split("_")
+        assert ca == "cat", "Categorical placeholder should start with cat_"
+        return self._choices[int(i)]
+
+    def get_ph(self):
+        return [f"cat_{i}" for i in range(len(self._choices))]
+
+
+class _Function_Value:
+    """Replaced value for function typed objects.
+    """
+    def __init__(self, fn):
+        self._fn = fn
+
+    def get(self, spec):
+        """Some functions take a resolved spec dict as input.
+        """
+        return self._fn.sample(spec=spec)
+
+    def get_ph(self):
+        return "fn_ph"
+
+
+class _Ref_Value:
+    """Replaced value for all other reference typed objects.
+    """
+    def __init__(self, value):
+        self._value = value
+
+    def get(self):
+        return self._value
+
+    def get_ph(self):
+        return "ref_ph"
+
+
+@DeveloperAPI
+def replace_references(spec: Any, replaced: Dict, prefix: Tuple = ()) -> Dict:
```

Comment:
Nit: We can name this `inject_placeholders`? Then it is consistent with `resolve_placeholders`. Also, "references" is not all that gets replaced.

Reply:
like it.